Nushell support

zaucy opened the pull request:
closes #463

Schniz commented on src/shell/nushell.rs:
```diff
@@ -0,0 +1,26 @@
+
+impl Shell for Nushell {
+    fn to_clap_shell(&self) -> clap_complete::Shell {
+        panic!("Shell completion is not supported for Nushell (yet.)");
```
Can we link to the issue in the Clap repo?
* https://github.com/clap-rs/clap/issues/2778

Schniz commented on src/shell/nushell.rs:
```diff
@@ -0,0 +1,26 @@
+        let path = path
+            .to_str()
+            .ok_or_else(|| anyhow::anyhow!("Can't convert path to string"))?;
+        Ok(format!("let-env PATH = ($env.PATH | prepend {:?})", path))
```
what's the difference between `let-env PATH` and `pathvar`?

* x-ref: https://github.com/Schniz/fnm/issues/463#issuecomment-1002651981

zaucy replied:
I don't know - my nushell installation doesn't have `pathvar`

```
Error: nu::shell::external_command (link)

  × External command
   ╭─[entry #3:1:1]
 1 │ pathvar
   · ───┬───
   ·    ╰── can't run executable
   ╰────
  help: No such file or directory (os error 2)
```

[The nushell docs recommend this way to add to your `PATH`](https://www.nushell.sh/book/configuration.html#path-configuration)

zaucy commented:
It seems to be working, however I cannot for the life of me figure out how to get the `warn_if_multishell_path_not_in_path_env_var` not to trigger in nushell. I printed out the `PATH` environment variable and it doesn't contain the hook ends.

I adapted the [starship install instructions](https://starship.rs/guide/#%F0%9F%9A%80-installation) for fnm like so.

Added to `$nu.env-path`
```nushell
mkdir ~/.cache/fnm
fnm env --use-on-cd | save ~/.cache/fnm/init.nu
```

Added to end of `$nu.config-path`
```nushell
source ~/.cache/fnm/init.nu
```

zaucy commented:
In case you're wondering - we can't eval the `fm env` AFAIK. I'm assuming this is why starship writes to a file as well.

https://www.nushell.sh/book/commands/source.html

This doesn't work

```nushell
Error: nu::parser::unexpected_keyword (link)

  × Statement used in pipeline.
   ╭─[entry #6:1:1]
 1 │ fnm env | source
   ·           ───┬──
   ·              ╰── not allowed in pipeline
   ╰────
  help: 'source' keyword is not allowed in pipeline. Use 'source' by itself, outside of a pipeline.


```

Schniz commented:
so when you invoke `fnm env` the output will be the file path:

```sh-session
$ fnm env --shell=nu
/tmp/fnm-env-MY_FILE-DATE_OR_TIME-PID.env
```

Schniz commented on src/shell/nushell.rs:
```diff
@@ -0,0 +1,53 @@
+                r#"
+                    let-env config = ($env.config | upsert hooks.env_change.PWD {{
+                        [
+                            {{
+                                condition: {{|_, after| ($after | path join .node-version | path exists) }}
+                                code: "fnm use --silent-if-unchanged"
+                            }}
```
do we need another one of these blocks for `.nvmrc`?

zaucy replied:
Ah I removed it temporarily while trying to get the syntax correct. I put it back here: https://github.com/Schniz/fnm/pull/801/commits/728013f45ba65dd7cee6306687725f1bf1085d6a

Schniz commented on src/shell/nushell.rs:
```diff
@@ -0,0 +1,53 @@
+        Ok(match config.version_file_strategy() {
+            VersionFileStrategy::Local => formatdoc!(
+                r#"
+                    let-env config = ($env.config | upsert hooks.env_change.PWD {{
```
brilliant!

Schniz reviewed (commented):
looks very good tbh. I'd love to know your input on the file creation suggestion to make initialization easier. It still does not integrate our e2e tests but this would be fairly hard to implement. Do you want to take a look at them? 👀 If it's too much, we can release this as experimental and I'll try to integrate the e2e tests when I have spare time

zaucy commented:
I can take a look at the e2e tests tomorrow, but no promises that I'll be able to :P

Schniz commented:
haha all good, you don't get paid so there are no expectations. what you've already done is appreciated already! :pray:

zaucy commented:
I gave the temp file technique a go and outputting the the generated file name and sourcing that is a no-go

Relevant post: https://github.com/nushell/nushell/discussions/4905

zaucy commented:
I added an option to write to a known file path instead. Nushell setup looks something like this now.

Add to your `$nu.config-path`
```nushell
fnm env --shell=nushell --use-on-cd --file
source ~/.fnm/fnm_env.nu
```

zaucy commented:
In light of the [newest nushell release notes](https://www.nushell.sh/blog/2022-09-06-nushell-0_68.html) I think this might be better suited as a module or overlay. I'm going to investigate over the next week.

jeroen-meijer commented:
Any way I can help with this? Just migrated to Nushell and love to see Rust projects grow.

zaucy commented:
Took another crack at finishing this up!

I removed the `--file` flag I added earlier. Now I have some hard set file paths exclusively for nushell. The nushell setup is pretty good now IMO. This is the output you get:

```sh
〉fnm env --use-on-cd                                                                                                                                                               
# Append to $nu.env-path
# source-env ~/.fnm/fnm_env.nu

# Append to $nu.config-path
# source ~/.fnm/fnm_config.nu

# For convenience:
echo "source-env ~/.fnm/fnm_env.nu" | save $nu.env-path --append
echo "source ~/.fnm/fnm_config.nu" | save $nu.config-path --append
```

mpizenberg commented:
Hi, are there still issues to solve for this PR? It seems like all was good 6 months ago so I'm just wondering what's the current fnm status in nushell as I'm interested in trying nushell and would love to continue using fnm there.

zaucy commented:
> Hi, are there still issues to solve for this PR? It seems like all was good 6 months ago so I'm just wondering what's the current fnm status in nushell as I'm interested in trying nushell and would love to continue using fnm there.

I'm not sure what is in this PR is acceptable since it behaves very different than all other shells. I've only been maintaining it for my local dev machines. #463 gives you some good ideas with the `--json` flag to integrate with nushell. Or you can install fnm from my branch! Be warned though I've only tested with nushell `0.78.0`.

caliguIa commented:
any updates on this? very much interested, thanks for all your work thus far!

vvuk commented:
In case anyone comes here and is looking for a manual solution -- `nushell 0.93`:

In `~/.config/nushell/env.nu` add:
```
use std "path add"
fnm env --json | from json | load-env
path add ($env.FNM_MULTISHELL_PATH + "/bin")
```

In `~/.config/nushell/config.nu` add:
```
$env.config.hooks.env_change.PWD = [...$env.config.hooks.env_change.PWD
    { |_, after|
      if (($after | path join .node-version | path exists) or ($after | path join .nvmrc | path exists)) {
          fnm use --silent-if-unchanged
      }
    }]
```

seems to work for me.

zophiana commented:
> In case anyone comes here and is looking for a manual solution -- `nushell 0.93`:
> 
> In `~/.config/nushell/env.nu` add:
> 
> ```nu
> use std "path add"
> fnm env --json | from json | load-env
> path add ($env.FNM_MULTISHELL_PATH + "/bin")
> ```

For me this did not work, but this fixed it -- `nushell 0.102`:

```nu
use std "path add"
fnm env --json | from json | load-env
path add $env.FNM_MULTISHELL_PATH
```

The rest stayed the same

2bndy5 commented on src/shell/nushell.rs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    fn to_clap_shell(&self) -> clap_complete::Shell {
+        panic!("Shell completion is not supported for Nushell (yet.) See https://github.com/clap-rs/clap/issues/2778 for updates.");
```
This info appears to be no longer accurate. The referenced issue (https://github.com/clap-rs/clap/issues/2778) was closed in favor of a new crate, which has since been transferred to the clap-rs/clap project: [clap_complete_nushell](https://github.com/clap-rs/clap/blob/master/clap_complete_nushell) [![Crates.io](https://img.shields.io/crates/v/clap_complete_nushell)](https://crates.io/crates/clap_complete_nushell)
